File: src/runner.ts
import { spawn } from "node:child_process";
import { promises as fs } from "node:fs";
import os from "node:os";
import path from "node:path";

import { compareReceipts } from "./compare.js";
import { sha256, sha256File, stableStringify } from "./hash.js";
import type {
  ArtifactReceipt,
  EnvironmentReceipt,
  IncompleteEvidence,
  NormalizerConfig,
  ProbeConfig,
  ProbeReceipt,
  RunReceipt,
  RuntimeIdentity,
  TheseusContract,
  VerificationResult,
} from "./types.js";

interface CommandResult {
  exitCode: number | null;
  signal: NodeJS.Signals | null;
  stdout: string;
  stderr: string;
  timedOut: boolean;
  durationMs: number;
  spawnError?: string;
}

interface PreparedArtifact extends ArtifactReceipt {
  tarballPath: string;
}

export async function verifyTarget(target: string, contract: TheseusContract): Promise<VerificationResult> {
  const tempRoot = await fs.mkdtemp(path.join(os.tmpdir(), "theseus-"));
  const receipts: RunReceipt[] = [];
  const incomplete: IncompleteEvidence[] = [];
  try {
    const npmVersionResult = await runCommand("npm", ["--version"], process.cwd(), 10_000);
    if (npmVersionResult.exitCode !== 0) throw new Error(`unable to resolve npm version: ${npmVersionResult.stderr}`);
    const environment: EnvironmentReceipt = {
      platform: process.platform,
      arch: process.arch,
      node: process.version,
      npm: npmVersionResult.stdout.trim(),
    };
    const artifact = await prepareArtifact(target, tempRoot);

    if (contract.source) {
      const sourceResult = await runSource(contract, environment);
      if (sourceResult.receipt) receipts.push(sourceResult.receipt);
      incomplete.push(...sourceResult.incomplete);
    }

    for (const [profileName, profile] of Object.entries(contract.profiles)) {
      const profileResult = await runInstalledProfile(profileName, profile.installArgs, artifact, contract, environment, tempRoot);
      if (profileResult.receipt) receipts.push(profileResult.receipt);
      incomplete.push(...profileResult.incomplete);
    }

    return {
      artifact: stripTarballPath(artifact),
      receipts,
      incomplete,
      comparison: compareReceipts(contract, receipts),
    };
  } finally {
    await fs.rm(tempRoot, { recursive: true, force: true });
  }
}

async function prepareArtifact(target: string, tempRoot: string): Promise<PreparedArtifact> {
  const artifactDir = path.join(tempRoot, "artifact");
  await fs.mkdir(artifactDir, { recursive: true });
  const resolvedTarget = target.startsWith(".") || target.startsWith("/") ? path.resolve(target) : target;
  const packed = await runCommand("npm", ["pack", resolvedTarget, "--json", "--pack-destination", artifactDir], process.cwd(), 120_000);
  if (packed.exitCode !== 0 || packed.timedOut || packed.spawnError) {
    const safeStderr = await sanitizeDiagnostic(packed.stderr, tempRoot);
    throw new Error(`npm pack failed: ${packed.spawnError ?? safeStderr.trim() ?? `exit ${String(packed.exitCode)}`}`);
  }
  let entries: unknown;
  try {
    entries = JSON.parse(packed.stdout);
  } catch {
    throw new Error(`npm pack did not return valid JSON: ${packed.stdout.slice(0, 400)}`);
  }
  if (!Array.isArray(entries) || entries.length !== 1) throw new Error("npm pack must return exactly one artifact");
  const entry = entries[0] as Record<string, unknown>;
  if (typeof entry.filename !== "string" || typeof entry.name !== "string" || typeof entry.version !== "string") {
    throw new Error("npm pack JSON is missing filename, name, or version");
  }
  const tarballPath = path.join(artifactDir, entry.filename);
  const stat = await fs.stat(tarballPath);
  return {
    packageName: entry.name,
    packageVersion: entry.version,
    filename: entry.filename,
    sha256: await sha256File(tarballPath),
    size: stat.size,
    tarballPath,
  };
}

async function runSource(
  contract: TheseusContract,
  environment: EnvironmentReceipt,
): Promise<{ receipt?: RunReceipt; incomplete: IncompleteEvidence[] }> {
  const source = contract.source!;
  const incomplete: IncompleteEvidence[] = [];
  const runtimeBase = {
    kind: "source" as const,
    binName: path.basename(source.command),
    executablePath: source.command,
    executableRealPath: source.command,
    optionalDependencies: [],
  };
  const canonical = stableStringify(runtimeBase);
  const runtime: RuntimeIdentity = { ...runtimeBase, canonical, sha256: sha256(canonical) };
  const sourceCwd = path.resolve(source.cwd);
  const sourceRealPath = await fs.realpath(sourceCwd);
  const probes = await runProbes(
    source.command,
    source.args,
    sourceCwd,
    contract.probes,
    contract.normalizers,
    "source",
    incomplete,
    [sourceCwd, sourceRealPath],
  );
  return {
    receipt: { subject: "source", profile: "source", installArgs: [], environment, runtime, probes },
    incomplete,
  };
}

async function runInstalledProfile(
  profileName: string,
  installArgs: string[],
  artifact: PreparedArtifact,
  contract: TheseusContract,
  environment: EnvironmentReceipt,
  tempRoot: string,
): Promise<{ receipt?: RunReceipt; incomplete: IncompleteEvidence[] }> {
  const incomplete: IncompleteEvidence[] = [];
  const projectDir = path.join(tempRoot, `profile-${safeName(profileName)}`);
  await fs.mkdir(projectDir, { recursive: true });
  await fs.writeFile(path.join(projectDir, "package.json"), JSON.stringify({ name: `theseus-consumer-${safeName(profileName)}`, private: true }), "utf8");
  const install = await runCommand(
    "npm",
    ["install", "--no-audit", "--no-fund", "--loglevel=error", ...installArgs, artifact.tarballPath],
    projectDir,
    180_000,
  );
  if (install.exitCode !== 0 || install.timedOut || install.spawnError) {
    const safeStderr = await sanitizeDiagnostic(install.stderr, tempRoot);
    incomplete.push({
      code: "THS-INCOMPLETE-001",
      profile: profileName,
      stage: "install",
      message: install.timedOut ? "npm install timed out" : install.spawnError ?? `npm install exited ${String(install.exitCode)}`,
      stderr: safeStderr,
    });
    return { incomplete };
  }

  const packageRoot = packageDirectory(path.join(projectDir, "node_modules"), artifact.packageName);
  let packageJson: Record<string, unknown>;
  try {
    packageJson = JSON.parse(await fs.readFile(path.join(packageRoot, "package.json"), "utf8")) as Record<string, unknown>;
  } catch (error) {
    incomplete.push({
      code: "THS-INCOMPLETE-001",
      profile: profileName,
      stage: "resolve-bin",
      message: `installed package.json could not be read: ${error instanceof Error ? error.message : String(error)}`,
    });
    return { incomplete };
  }

  let binName: string;
  try {
    binName = resolveBinName(packageJson, artifact.packageName, contract.bin);
  } catch (error) {
    incomplete.push({
      code: "THS-INCOMPLETE-001",
      profile: profileName,
      stage: "resolve-bin",
      message: error instanceof Error ? error.message : String(error),
    });
    return { incomplete };
  }
  const binFile = process.platform === "win32" ? `${binName}.cmd` : binName;
  const executable = path.join(projectDir, "node_modules", ".bin", binFile);
  let realExecutable: string;
  try {
    realExecutable = await fs.realpath(executable);
  } catch (error) {
    incomplete.push({
      code: "THS-INCOMPLETE-001",
      profile: profileName,
      stage: "resolve-bin",
      message: `bin ${binName} is missing: ${error instanceof Error ? error.message : String(error)}`,
    });
    return { incomplete };
  }

  const optionalDependencies = await installedOptionalDependencies(projectDir, packageJson);
  const projectRealPath = await fs.realpath(projectDir);
  const runtimeBase = {
    kind: "installed" as const,
    packageName: artifact.packageName,
    packageVersion: artifact.packageVersion,
    binName,
    executablePath: relative(projectDir, executable),
    executableRealPath: relative(projectRealPath, realExecutable),
    optionalDependencies,
  };
  const canonical = stableStringify(runtimeBase);
  const runtime: RuntimeIdentity = { ...runtimeBase, canonical, sha256: sha256(canonical) };
  const probes = await runProbes(
    executable,
    [],
    projectDir,
    contract.probes,
    contract.normalizers,
    profileName,
    incomplete,
    [projectDir, projectRealPath],
  );
  return {
    receipt: { subject: "installed", profile: profileName, installArgs, environment, runtime, probes },
    incomplete,
  };
}

async function runProbes(
  command: string,
  prefixArgs: string[],
  cwd: string,
  probes: ProbeConfig[],
  normalizers: NormalizerConfig[],
  profile: string,
  incomplete: IncompleteEvidence[],
  automaticRoots: string[],
): Promise<ProbeReceipt[]> {
  const receipts: ProbeReceipt[] = [];
  for (const probe of probes) {
    const result = await runCommand(command, [...prefixArgs, ...probe.argv], cwd, probe.timeoutMs);
    const stdout = normalize(result.stdout, "stdout", probe.id, normalizers, automaticRoots);
    const stderr = normalize(result.stderr, "stderr", probe.id, normalizers, automaticRoots);
    receipts.push({
      id: probe.id,
      argv: probe.argv,
      exitCode: result.exitCode,
      signal: result.signal,
      timedOut: result.timedOut,
      stdout,
      stderr,
      stdoutSha256: sha256(stdout),
      stderrSha256: sha256(stderr),
      durationMs: result.durationMs,
    });
    if (result.timedOut || result.spawnError) {
      const evidence: IncompleteEvidence = {
        code: "THS-INCOMPLETE-001",
        profile,
        stage: "probe",
        probe: probe.id,
        message: result.timedOut ? `probe timed out after ${probe.timeoutMs}ms` : result.spawnError!,
      };
      if (stderr) evidence.stderr = stderr;
      incomplete.push(evidence);
    }
  }
  return receipts;
}

function normalize(
  value: string,
  field: "stdout" | "stderr",
  probe: string,
  normalizers: NormalizerConfig[],
  automaticRoots: string[],
): string {
  let normalized = value.replaceAll("\r\n", "\n");
  for (const root of [...automaticRoots].sort((left, right) => right.length - left.length)) {
    normalized = normalized.replaceAll(root, "<subject-root>");
  }
  for (const config of normalizers) {
    if (config.probe && config.probe !== probe) continue;
    if (!config.fields.includes(field)) continue;
    normalized = normalized.replace(new RegExp(config.pattern, config.flags), config.replacement);
  }
  return normalized;
}

async function sanitizeDiagnostic(value: string, tempRoot: string): Promise<string> {
  let sanitized = value.replaceAll("\r\n", "\n");
  const realTempRoot = await fs.realpath(tempRoot);
  for (const root of [tempRoot, realTempRoot].sort((left, right) => right.length - left.length)) {
    sanitized = sanitized.replaceAll(root, "<theseus-temp>");
  }
  return sanitized.replace(/(?:\/[^\s]+)?\/\.npm\/_logs\/[^\s]+/g, "<npm-log>");
}

function resolveBinName(packageJson: Record<string, unknown>, packageName: string, requested?: string): string {
  const bin = packageJson.bin;
  let entries: Array<[string, string]>;
  if (typeof bin === "string") {
    entries = [[packageName.split("/").at(-1)!, bin]];
  } else if (bin !== null && typeof bin === "object" && !Array.isArray(bin)) {
    entries = Object.entries(bin as Record<string, unknown>).filter((entry): entry is [string, string] => typeof entry[1] === "string");
  } else {
    entries = [];
  }
  if (requested) {
    if (!entries.some(([name]) => name === requested)) throw new Error(`requested bin ${requested} is not declared by ${packageName}`);
    return requested;
  }
  if (entries.length !== 1) throw new Error(`${packageName} declares ${entries.length} bins; contract.bin is required`);
  return entries[0]![0];
}

async function installedOptionalDependencies(projectDir: string, packageJson: Record<string, unknown>): Promise<Array<{ name: string; version: string }>> {
  const raw = packageJson.optionalDependencies;
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return [];
  const results: Array<{ name: string; version: string }> = [];
  for (const name of Object.keys(raw as Record<string, unknown>).sort()) {
    try {
      const manifest = JSON.parse(await fs.readFile(path.join(packageDirectory(path.join(projectDir, "node_modules"), name), "package.json"), "utf8")) as Record<string, unknown>;
      if (typeof manifest.version === "string") results.push({ name, version: manifest.version });
    } catch {
      // Absence is the evidence for an omitted or unsupported optional dependency.
    }
  }
  return results;
}

function packageDirectory(nodeModules: string, name: string): string {
  return path.join(nodeModules, ...name.split("/"));
}

function relative(root: string, target: string): string {
  return path.relative(root, target).split(path.sep).join("/");
}

function safeName(value: string): string {
  return value.replace(/[^A-Za-z0-9_-]/g, "-");
}

function stripTarballPath(artifact: PreparedArtifact): ArtifactReceipt {
  const { tarballPath: _tarballPath, ...receipt } = artifact;
  return receipt;
}

function runCommand(command: string, args: string[], cwd: string, timeoutMs: number): Promise<CommandResult> {
  return new Promise((resolve) => {
    const started = Date.now();
    let stdout = "";
    let stderr = "";
    let timedOut = false;
    let spawnError: string | undefined;
    const detached = process.platform !== "win32";
    const child = spawn(command, args, {
      cwd,
      env: process.env,
      detached,
      shell: process.platform === "win32",
      stdio: ["ignore", "pipe", "pipe"],
    });
    child.stdout.setEncoding("utf8");
    child.stderr.setEncoding("utf8");
    child.stdout.on("data", (chunk: string) => (stdout += chunk));
    child.stderr.on("data", (chunk: string) => (stderr += chunk));
    child.on("error", (error) => {
      spawnError = error.message;
    });
    const timer = setTimeout(() => {
      timedOut = true;
      if (detached && child.pid !== undefined) {
        try {
          process.kill(-child.pid, "SIGKILL");
        } catch {
          child.kill("SIGKILL");
        }
      } else {
        child.kill("SIGKILL");
      }
    }, timeoutMs);
    child.on("close", (exitCode, signal) => {
      clearTimeout(timer);
      const result: CommandResult = {
        exitCode,
        signal,
        stdout,
        stderr,
        timedOut,
        durationMs: Date.now() - started,
      };
      if (spawnError) result.spawnError = spawnError;
      resolve(result);
    });
  });
}

File: internal/runner/runner.go
package runner

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"time"

	"github.com/JeremyDev87/theseus/internal/canonical"
	"github.com/JeremyDev87/theseus/internal/compare"
	"github.com/JeremyDev87/theseus/internal/contract"
	"github.com/JeremyDev87/theseus/internal/model"
	processrun "github.com/JeremyDev87/theseus/internal/process"
	"github.com/JeremyDev87/theseus/internal/runtimeidentity"
)

type preparedArtifact struct {
	model.ArtifactReceipt
	TarballPath string
}

type packageManifest struct {
	Name                 string                     `json:"name"`
	Version              string                     `json:"version"`
	Bin                  json.RawMessage            `json:"bin"`
	OptionalDependencies map[string]json.RawMessage `json:"optionalDependencies"`
}

var (
	currentWorkingDirectory = os.Getwd
	removeTemporaryRoot     = os.RemoveAll
)

func Verify(target string, config model.Contract) (result model.VerificationResult, err error) {
	tempRoot, err := os.MkdirTemp("", "theseus-")
	if err != nil {
		return model.VerificationResult{}, err
	}
	defer func() {
		if cleanupErr := removeTemporaryRoot(tempRoot); cleanupErr != nil {
			result = model.VerificationResult{}
			err = errors.Join(err, fmt.Errorf("unable to remove temporary root: %w", cleanupErr))
		}
	}()
	cwd, err := currentWorkingDirectory()
	if err != nil {
		return model.VerificationResult{}, fmt.Errorf("unable to resolve working directory: %w", err)
	}
	nodeResult := processrun.Run("node", []string{"--version"}, cwd, 10*time.Second)
	if nodeResult.ExitCode == nil || *nodeResult.ExitCode != 0 || nodeResult.SpawnError != "" {
		return model.VerificationResult{}, fmt.Errorf("unable to resolve node version: %s", commandDetail(nodeResult))
	}
	npmResult := processrun.Run("npm", []string{"--version"}, cwd, 10*time.Second)
	if npmResult.ExitCode == nil || *npmResult.ExitCode != 0 || npmResult.SpawnError != "" {
		return model.VerificationResult{}, fmt.Errorf("unable to resolve npm version: %s", commandDetail(npmResult))
	}
	environment := model.EnvironmentReceipt{Platform: nodePlatform(), Arch: nodeArch(), Node: strings.TrimSpace(nodeResult.Stdout), NPM: strings.TrimSpace(npmResult.Stdout)}
	artifact, err := prepareArtifact(target, tempRoot, cwd)
	if err != nil {
		return model.VerificationResult{}, err
	}
	receipts := []model.RunReceipt{}
	incomplete := []model.IncompleteEvidence{}
	if config.Source != nil {
		receipt, evidence := runSource(config, environment)
		if receipt != nil {
			receipts = append(receipts, *receipt)
		}
		incomplete = append(incomplete, evidence...)
	}
	for _, profileName := range config.ProfileOrder {
		profile := config.Profiles[profileName]
		receipt, evidence := runProfile(profileName, profile.InstallArgs, artifact, config, environment, tempRoot)
		if receipt != nil {
			receipts = append(receipts, *receipt)
		}
		incomplete = append(incomplete, evidence...)
	}
	return model.VerificationResult{
		Artifact:   artifact.ArtifactReceipt,
		Receipts:   receipts,
		Incomplete: incomplete,
		Comparison: compare.Receipts(config, receipts),
	}, nil
}

func prepareArtifact(target, tempRoot, cwd string) (preparedArtifact, error) {
	artifactDir := filepath.Join(tempRoot, "artifact")
	if err := os.MkdirAll(artifactDir, 0o755); err != nil {
		return preparedArtifact{}, err
	}
	resolvedTarget := target
	if strings.HasPrefix(target, ".") || filepath.IsAbs(target) {
		var err error
		resolvedTarget, err = filepath.Abs(target)
		if err != nil {
			return preparedArtifact{}, err
		}
	}
	result := processrun.Run("npm", []string{"pack", resolvedTarget, "--json", "--pack-destination", artifactDir}, cwd, 120*time.Second)
	if result.ExitCode == nil || *result.ExitCode != 0 || result.TimedOut || result.SpawnError != "" {
		return preparedArtifact{}, fmt.Errorf("npm pack failed: %s", sanitizeDiagnostic(commandDetail(result), tempRoot))
	}
	var entries []struct {
		Filename string `json:"filename"`
		Name     string `json:"name"`
		Version  string `json:"version"`
	}
	if err := json.Unmarshal([]byte(result.Stdout), &entries); err != nil {
		value := result.Stdout
		if len(value) > 400 {
			value = value[:400]
		}
		return preparedArtifact{}, fmt.Errorf("npm pack did not return valid JSON: %s", value)
	}
	if len(entries) != 1 {
		return preparedArtifact{}, fmt.Errorf("npm pack must return exactly one artifact")
	}
	entry := entries[0]
	if entry.Filename == "" || entry.Name == "" || entry.Version == "" {
		return preparedArtifact{}, fmt.Errorf("npm pack JSON is missing filename, name, or version")
	}
	tarballPath := filepath.Join(artifactDir, entry.Filename)
	info, err := os.Stat(tarballPath)
	if err != nil {
		return preparedArtifact{}, err
	}
	digest, err := canonical.SHA256File(tarballPath)
	if err != nil {
		return preparedArtifact{}, err
	}
	return preparedArtifact{ArtifactReceipt: model.ArtifactReceipt{PackageName: entry.Name, PackageVersion: entry.Version, Filename: entry.Filename, SHA256: digest, Size: info.Size()}, TarballPath: tarballPath}, nil
}

func runSource(config model.Contract, environment model.EnvironmentReceipt) (*model.RunReceipt, []model.IncompleteEvidence) {
	source := config.Source
	cwd, err := filepath.Abs(source.CWD)
	if err != nil {
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: "source", Stage: "probe", Message: err.Error()}}
	}
	realCWD, err := filepath.EvalSymlinks(cwd)
	if err != nil {
		realCWD = cwd
	}
	runtimeIdentity, err := runtimeidentity.Create(model.RuntimeIdentity{
		Kind: "source", BinName: filepath.Base(source.Command), ExecutablePath: source.Command, ExecutableRealPath: source.Command,
		OptionalDependencies: []model.OptionalDependencyReceipt{},
	})
	if err != nil {
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: "source", Stage: "probe", Message: err.Error()}}
	}
	incomplete := []model.IncompleteEvidence{}
	probes := runProbes(source.Command, source.Args, cwd, config, "source", &incomplete, []string{cwd, realCWD})
	return &model.RunReceipt{Subject: "source", Profile: "source", InstallArgs: []string{}, Environment: environment, Runtime: runtimeIdentity, Probes: probes}, incomplete
}

func runProfile(profileName string, installArgs []string, artifact preparedArtifact, config model.Contract, environment model.EnvironmentReceipt, tempRoot string) (*model.RunReceipt, []model.IncompleteEvidence) {
	incomplete := []model.IncompleteEvidence{}
	projectDir := filepath.Join(tempRoot, "profile-"+safeName(profileName))
	if err := os.MkdirAll(projectDir, 0o755); err != nil {
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: profileName, Stage: "install", Message: err.Error()}}
	}
	manifestBytes, _ := json.Marshal(map[string]any{"name": "theseus-consumer-" + safeName(profileName), "private": true})
	if err := os.WriteFile(filepath.Join(projectDir, "package.json"), manifestBytes, 0o644); err != nil {
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: profileName, Stage: "install", Message: err.Error()}}
	}
	args := []string{"install", "--no-audit", "--no-fund", "--loglevel=error"}
	args = append(args, installArgs...)
	args = append(args, artifact.TarballPath)
	install := processrun.Run("npm", args, projectDir, 180*time.Second)
	if install.ExitCode == nil || *install.ExitCode != 0 || install.TimedOut || install.SpawnError != "" {
		message := fmt.Sprintf("npm install exited %s", displayExit(install.ExitCode))
		if install.TimedOut {
			message = "npm install timed out"
		} else if install.SpawnError != "" {
			message = install.SpawnError
		}
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: profileName, Stage: "install", Message: message, Stderr: sanitizeDiagnostic(install.Stderr, tempRoot)}}
	}
	packageRoot := packageDirectory(filepath.Join(projectDir, "node_modules"), artifact.PackageName)
	manifest, err := readManifest(filepath.Join(packageRoot, "package.json"))
	if err != nil {
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: profileName, Stage: "resolve-bin", Message: "installed package.json could not be read: " + err.Error()}}
	}
	binName, err := resolveBin(manifest, artifact.PackageName, config.Bin)
	if err != nil {
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: profileName, Stage: "resolve-bin", Message: err.Error()}}
	}
	binFile := binName
	if runtime.GOOS == "windows" {
		binFile += ".cmd"
	}
	executable := filepath.Join(projectDir, "node_modules", ".bin", binFile)
	realExecutable, err := filepath.EvalSymlinks(executable)
	if err != nil {
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: profileName, Stage: "resolve-bin", Message: fmt.Sprintf("bin %s is missing: %s", binName, err)}}
	}
	optionalDependencies := installedOptionalDependencies(projectDir, manifest)
	projectRealPath, err := filepath.EvalSymlinks(projectDir)
	if err != nil {
		projectRealPath = projectDir
	}
	executablePath, _ := filepath.Rel(projectDir, executable)
	executableRealPath, _ := filepath.Rel(projectRealPath, realExecutable)
	executablePath = filepath.ToSlash(executablePath)
	executableRealPath = filepath.ToSlash(executableRealPath)
	runtimeIdentity, err := runtimeidentity.Create(model.RuntimeIdentity{
		Kind: "installed", PackageName: artifact.PackageName, PackageVersion: artifact.PackageVersion, BinName: binName,
		ExecutablePath: executablePath, ExecutableRealPath: executableRealPath, OptionalDependencies: optionalDependencies,
	})
	if err != nil {
		return nil, []model.IncompleteEvidence{{Code: "THS-INCOMPLETE-001", Profile: profileName, Stage: "resolve-bin", Message: err.Error()}}
	}
	probes := runProbes(executable, nil, projectDir, config, profileName, &incomplete, []string{projectDir, projectRealPath})
	return &model.RunReceipt{Subject: "installed", Profile: profileName, InstallArgs: append([]string{}, installArgs...), Environment: environment, Runtime: runtimeIdentity, Probes: probes}, incomplete
}

func runProbes(command string, prefixArgs []string, cwd string, config model.Contract, profile string, incomplete *[]model.IncompleteEvidence, automaticRoots []string) []model.ProbeReceipt {
	receipts := []model.ProbeReceipt{}
	for _, probe := range config.Probes {
		args := append(append([]string{}, prefixArgs...), probe.Argv...)
		result := processrun.Run(command, args, cwd, time.Duration(probe.TimeoutMS)*time.Millisecond)
		stdout := normalize(result.Stdout, "stdout", probe.ID, config.Normalizers, automaticRoots)
		stderr := normalize(result.Stderr, "stderr", probe.ID, config.Normalizers, automaticRoots)
		receipts = append(receipts, model.ProbeReceipt{ID: probe.ID, Argv: append([]string{}, probe.Argv...), ExitCode: result.ExitCode, Signal: result.Signal, TimedOut: result.TimedOut, Stdout: stdout, Stderr: stderr, StdoutSHA256: canonical.SHA256String(stdout), StderrSHA256: canonical.SHA256String(stderr), DurationMS: result.DurationMS})
		if result.TimedOut || result.SpawnError != "" {
			message := result.SpawnError
			if result.TimedOut {
				message = fmt.Sprintf("probe timed out after %dms", probe.TimeoutMS)
				if result.SpawnError != "" {
					message += "; cleanup error: " + result.SpawnError
				}
			}
			evidence := model.IncompleteEvidence{Code: "THS-INCOMPLETE-001", Profile: profile, Stage: "probe", Probe: probe.ID, Message: message}
			if stderr != "" {
				evidence.Stderr = stderr
			}
			*incomplete = append(*incomplete, evidence)
		}
	}
	return receipts
}

func normalize(value, field, probe string, configs []model.NormalizerConfig, automaticRoots []string) string {
	value = strings.ReplaceAll(value, "\r\n", "\n")
	roots := append([]string{}, automaticRoots...)
	sort.Slice(roots, func(left, right int) bool { return len(roots[left]) > len(roots[right]) })
	for _, root := range roots {
		value = strings.ReplaceAll(value, root, "<subject-root>")
	}
	for _, config := range configs {
		if config.Probe != "" && config.Probe != probe {
			continue
		}
		if !contains(config.Fields, field) {
			continue
		}
		prefix, _ := contract.RegexPrefix(config.Flags)
		expression := regexp.MustCompile(prefix + config.Pattern)
		if contract.HasGlobal(config.Flags) {
			value = expression.ReplaceAllString(value, config.Replacement)
		} else if match := expression.FindStringSubmatchIndex(value); match != nil {
			replacement := expression.ExpandString(nil, config.Replacement, value, match)
			value = value[:match[0]] + string(replacement) + value[match[1]:]
		}
	}
	return value
}

func sanitizeDiagnostic(value, tempRoot string) string {
	value = strings.ReplaceAll(value, "\r\n", "\n")
	roots := []string{tempRoot}
	if real, err := filepath.EvalSymlinks(tempRoot); err == nil {
		roots = append(roots, real)
	}
	sort.Slice(roots, func(left, right int) bool { return len(roots[left]) > len(roots[right]) })
	for _, root := range roots {
		value = strings.ReplaceAll(value, root, "<theseus-temp>")
	}
	unixLogs := regexp.MustCompile(`(?:/[^\s]+)?/\.npm/_logs/[^\s]+`)
	value = unixLogs.ReplaceAllString(value, "<npm-log>")
	windowsLogs := regexp.MustCompile(`(?i)(?:[A-Z]:\\|\\\\)[^\r\n]*?\\(?:npm-cache|\.npm)\\_logs\\[^\s]+`)
	return windowsLogs.ReplaceAllString(value, "<npm-log>")
}

func readManifest(path string) (packageManifest, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return packageManifest{}, err
	}
	var manifest packageManifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		return packageManifest{}, err
	}
	return manifest, nil
}

func resolveBin(manifest packageManifest, packageName, requested string) (string, error) {
	entries := map[string]string{}
	if len(manifest.Bin) > 0 && string(manifest.Bin) != "null" {
		var single string
		if err := json.Unmarshal(manifest.Bin, &single); err == nil {
			parts := strings.Split(packageName, "/")
			entries[parts[len(parts)-1]] = single
		} else {
			var values map[string]any
			if err := json.Unmarshal(manifest.Bin, &values); err == nil {
				for name, value := range values {
					if path, ok := value.(string); ok {
						entries[name] = path
					}
				}
			}
		}
	}
	if requested != "" {
		if _, exists := entries[requested]; !exists {
			return "", fmt.Errorf("requested bin %s is not declared by %s", requested, packageName)
		}
		return requested, nil
	}
	if len(entries) != 1 {
		return "", fmt.Errorf("%s declares %d bins; contract.bin is required", packageName, len(entries))
	}
	for name := range entries {
		return name, nil
	}
	return "", fmt.Errorf("internal: bin resolution found no entries for %s", packageName)
}

func installedOptionalDependencies(projectDir string, manifest packageManifest) []model.OptionalDependencyReceipt {
	names := []string{}
	for name := range manifest.OptionalDependencies {
		names = append(names, name)
	}
	sort.Strings(names)
	result := []model.OptionalDependencyReceipt{}
	for _, name := range names {
		dependency, err := readManifest(filepath.Join(packageDirectory(filepath.Join(projectDir, "node_modules"), name), "package.json"))
		if err == nil && dependency.Version != "" {
			result = append(result, model.OptionalDependencyReceipt{Name: name, Version: dependency.Version})
		}
	}
	return result
}

func packageDirectory(nodeModules, name string) string {
	parts := strings.Split(name, "/")
	return filepath.Join(append([]string{nodeModules}, parts...)...)
}
func safeName(value string) string {
	return regexp.MustCompile(`[^A-Za-z0-9_-]`).ReplaceAllString(value, "-")
}
func contains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}
func nodePlatform() string {
	if runtime.GOOS == "windows" {
		return "win32"
	}
	return runtime.GOOS
}
func nodeArch() string {
	switch runtime.GOARCH {
	case "amd64":
		return "x64"
	case "386":
		return "ia32"
	default:
		return runtime.GOARCH
	}
}
func displayExit(value *int) string {
	if value == nil {
		return "null"
	}
	return fmt.Sprint(*value)
}
func commandDetail(result processrun.Result) string {
	if result.SpawnError != "" {
		return result.SpawnError
	}
	if strings.TrimSpace(result.Stderr) != "" {
		return strings.TrimSpace(result.Stderr)
	}
	return "exit " + displayExit(result.ExitCode)
}
